File: Rust/light_follower.rs
/* 
Light Follower Robot NXC example
Author: David Laurent Reinhardt
Course: "System Design Projekt", Winter Semester 2025/2026

Description:
This program implements a simple light-following robot using a proportional controller.
The robot uses two light sensors to detect the intensity of light on its left and right sides.
Based on the difference in light intensity, it adjusts its motor speed to turn towards the light source.
*/

extern crate ev3dev_lang_rust;
use ev3dev_lang_rust::motors::{LargeMotor, MotorPort};
use ev3dev_lang_rust::sensors::{LightSensor, SensorPort};
use ev3dev_lang_rust::Ev3Result;


//  p(roportional)-controller
fn p_controller(sensor_left: &LightSensor, sensor_right: &LightSensor, p_term: f32) -> Ev3Result<i32> {
    let difference: i32 = sensor_left.value(0)? - sensor_right.value(0)? as i32;
    Ok((p_term * difference as f32) as i32)
}


//  "task main()" is executed when the program starts
fn main() -> Ev3Result<()> {

    // Constants
    let LIGHT_LEFT = LightSensor::new(SensorPort::In1)?;
    let LIGHT_RIGHT = LightSensor::new(SensorPort::In2)?;
    let MOTOR = LargeMotor::new(MotorPort::OutA)?;
    let P_TERM: f32 = 1.0;

    // variable to store p_controller output
    let mut motor_power: i32;

    //  loops forever
    loop {
        motor_power = p_controller(&LIGHT_LEFT, &LIGHT_RIGHT, P_TERM)?;
        MOTOR.set_speed(motor_power)?;
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
}

File: Python/light_follower.py
#!/usr/bin/env python3

"""
Light Follower Robot Python example
Author: David Laurent Reinhardt
Course: "System Design Projekt", Winter Semester 2025/2026

Description:
This program implements a simple light-following robot using a proportional controller.
The robot uses two light sensors to detect the intensity of light on its left and right sides.
Based on the difference in light intensity, it adjusts its motor speed to turn towards the light source.
"""


from time import sleep
from ev3dev2.motor import OUTPUT_A, LargeMotor
from ev3dev2.sensor import INPUT_1, INPUT_2
from ev3dev2.sensor.lego import LightSensor


# Constants
LIGHT_LEFT = LightSensor(INPUT_1)
LIGHT_RIGHT = LightSensor(INPUT_2)
MOTOR = LargeMotor(OUTPUT_A)
P_TERM: float = 1.0


# p(roportional)-controller function
def p_controller(sensor_left: int, sensor_right: int, p_term: float) -> float:
    difference: float = sensor_left - sensor_right
    return p_term * difference


# is executed when the program is started
if __name__ == "__main__":

    # variable to store p_controller output
    motor_power: float

    # loops forever
    while True:
        
        motor_power = p_controller(LIGHT_LEFT.reflected_light_intensity, LIGHT_RIGHT.reflected_light_intensity, P_TERM)
        MOTOR.on(motor_power)

        # wait for 100 milliseconds
        sleep(0.1)